src/middlewared/middlewared/plugins/pool.py

```diff
@@ -2950,6 +2907,38 @@ def add(attr):
                 numdisks = num
         return '%dK' % 2 ** ((numdisks * 4) - 1).bit_length()
 
+    @item_method
+    @accepts(Str('id', required=True))
+    async def attachments(self, oid):
+        """
+        Return a list of services dependent of this dataset.
+
+        Responsible for telling the user whether there is a related
+        share, asking for confirmation.
+        """
+        result = []
+        dataset = await self._get_instance(oid)
+        path = self.__attachments_path(dataset)
+        if path:
+            for delegate in self.attachment_delegates:
+                attachments = {"type": delegate.title, "attachments": []}
+                for attachment in await delegate.query(path, True):
+                    attachments["attachments"].append(await delegate.get_attachment_name(attachment))
+                if attachments["attachments"]:
+                    result.append(attachments)
+        return result
+
+    def __attachments_path(self, dataset):
+        if dataset['type'] == 'FILESYSTEM':
+            return dataset['mountpoint']
+
+        if dataset['type'] == 'VOLUME':
```

Comment:
If its a VOLUME its a zvol, I am confused why you're returning it as a pool?

Reply:
To make datasets and zvols share hierarchy.

This way, if I delete dataset `tank/ds1`, iSCSI extent `tank/ds1/ds2/zvol1` will also be deleted.

Otherwise I would have to check if we are deleting dataset or zvol in ISCSIFSAttachmentDelegate. And other delegates will be called with paths like `zvol/tank/ds1/ds2/zvol1` which does not have any sense.